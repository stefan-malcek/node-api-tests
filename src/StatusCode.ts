export const Ok = 200;
export const BadRequest = 400;
export const Unauthorized = 401;
export const Forbidden = 403;
export const NotFound = 404;
export const InternalError = 500;