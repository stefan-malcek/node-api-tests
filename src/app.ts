import {defaultMetadataStorage} from 'class-transformer/storage';
import {getFromContainer, MetadataStorage} from 'class-validator';
import {validationMetadatasToSchemas} from 'class-validator-jsonschema';
import express from 'express';
import {Action, getMetadataArgsStorage, useExpressServer} from 'routing-controllers';
import {routingControllersToSpec} from 'routing-controllers-openapi';
import swaggerUiExpress from 'swagger-ui-express';
import {BookController} from './controllers/BookController';
import {GenreController} from './controllers/GenreController';
import {ErrorHandlingMiddleware} from "./middlewares/ErrorHandlingMiddleware";
import JwtManager from "./services/JwtManager";

const authorizationChecker = async (action: Action, roles: string[]) => {
    const { authorization } = action.request.headers;
    if (!authorization) {
        return false;
    }

    const jwtToken = authorization.replace('Bearer ', '');
    const token = JwtManager.decodeToken(jwtToken);
    if (!token) {
        return false;
    }

    // You should verify userId.

    if (!roles.length) {
        return true;
    }

   return roles.some(role => token.role === role);
};

export const run = async (callback: (app: any) => void) => {
    const application = express();

    application.get('/', (_req, res) => res.send('API works!'));

    // creates express app, registers all controller routes
    const routingControllersOptions = {
        authorizationChecker,
        controllers: [GenreController, BookController],
        middlewares: [ErrorHandlingMiddleware],
        routePrefix: '/api',
        classTransformer: true,
        defaultErrorHandler: false,
        cors: true,
    };
    useExpressServer(application, routingControllersOptions);

    const metadatas = (getFromContainer(MetadataStorage) as any)
        .validationMetadatas;

    const schemas = validationMetadatasToSchemas(metadatas, {
        classTransformerMetadataStorage: defaultMetadataStorage,
        refPointerPrefix: '#/components/schemas/',
    });

    const storage = getMetadataArgsStorage();
    const spec = routingControllersToSpec(storage, routingControllersOptions, {
        components: {
            schemas,
            securitySchemes: {
                bearerAuth: {
                    scheme: 'bearer',
                    type: 'http',
                },
            },
        },
        title: 'Book Shelf API',
        version: 'v1',
    });

    application.use('/swagger', swaggerUiExpress.serve, swaggerUiExpress.setup(spec));

    // Init application here.

    callback(application);
};

export const dispose = async () => {
    // Clean resource here, e.g.: database connection
};
