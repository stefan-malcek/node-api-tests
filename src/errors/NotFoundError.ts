export class NotFoundError extends Error {
  constructor(entity: string, id: any) {
    super(`Entity ${entity} (${id}) was not found.`);
  }
}
