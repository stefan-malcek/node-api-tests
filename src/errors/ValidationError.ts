import {BusinessError, ValdiationFailed} from "./index";

export class ValidationError extends Error {
    public description: string;

    constructor(error: BusinessError) {
        super(ValdiationFailed.description);
        this.name = error.name;
        this.description = error.description;
    }
}
