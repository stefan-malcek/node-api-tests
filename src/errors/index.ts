export interface BusinessError {
    name: string;
    description: string
}

export const ValidationFailed : BusinessError = {
    name: 'VALIDATION_FAILED',
    description: 'Validation failed, see \'errors\' for further details.'
}

export const InvalidCredentials: BusinessError = {
    name: 'INVALID_CREDENTIALS',
    description: 'Username or password is invalid.'
}

export const InvalidGender: BusinessError = {name: 'INVALID_GENDER', description: "Gender doesn't exist."}
export const IsbnNotUnique: BusinessError = {name: 'ISBN_NOT_UNIQUE', description: 'ISBN already exists.'}