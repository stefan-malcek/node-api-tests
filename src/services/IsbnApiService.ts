import axios from 'axios';

export class IsbnApiService {
    // Should be in config.
    private readonly baseUrl = 'https://api.altmetric.com/v1/isbn';

    public async getBookInfo(isbn: string) {
        try {
            const response = await axios.get(`${this.baseUrl}/${isbn}`);
            return response.data;
        } catch (e) {
            console.log(e);
            return null;
        }
    }
}