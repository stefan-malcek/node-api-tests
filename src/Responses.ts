import {Type} from 'class-transformer';
import {
    IsInt,
    IsString,
    ValidateNested,
} from 'class-validator';
import * as StatusCode from './StatusCode';

export class ApiResponse {
    @IsInt()
    public statusCode: number;

    @IsString()
    public message: string;

    constructor(statusCode: number = StatusCode.Ok, message: string = "Ok") {
        this.statusCode = statusCode;
        this.message = message;
    }
}

export class ApiErrorResponse extends ApiResponse {
    @IsString()
    public name: string;

    @ValidateNested({each: true})
    @Type(() => String)
    public errors: string[];

    constructor(name: string, errors: string[]) {
        super(StatusCode.BadRequest, 'Validation failed, see \'errors\' for further details.');
        this.name = name;
        this.errors = errors;
    }
}
