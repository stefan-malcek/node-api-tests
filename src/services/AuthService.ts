import {LoginCommand, UserDto, UserProfileDto} from "../models/AuthModel";
import {ValidationError} from "../errors/ValidationError";
import {InvalidCredentials} from "../errors";
import JwtManager from "./JwtManager";

const users = [
    {id: 1, username: 'user', role: 'USER'},
    {id: 2, username: 'admin', role: 'ADMIN'}
];

export class AuthService {
    public login(login: LoginCommand): UserProfileDto {
        const user = users.find(user => user.username === login.username);
        if (!user) {
            throw new ValidationError(InvalidCredentials);
        }

        // Omit password verification.

        const profile = new UserProfileDto();

        const userDto = new UserDto();
        userDto.id = user.id;
        userDto.username = user.username;
        profile.user = user;

        const token = JwtManager.createToken(user.id, user.role);
        profile.token = token;

        return profile;
    }
}
