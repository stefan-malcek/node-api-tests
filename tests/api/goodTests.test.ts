import * as StatusCode from '../../src/StatusCode';
import {API_START_TIMEOUT} from '../constants';
import {TestApiClient} from '../TestApiClient';
import {InvalidGender, IsbnNotUnique, ValidationFailed} from '../../src/errors';
import {generateString} from './utils';
import {bookQueryFactory, createBookFactory, mockedIsbnApiResponse, updateBookFactory} from './factories/bookFactory';
import {IsbnApiService} from "../../src/services/IsbnApiService";

let client: TestApiClient;

const BASE_URL = '/api/books';
const QUERY_BOOK_ID = 3;
const EDIT_BOOK_ID = 4;
const DELETE_BOOK_ID = 2;
const INVALID_BOOK_ID = 565;

beforeAll(async done => {
    client = new TestApiClient(BASE_URL);
    client.resetNow();

    await client.runServer();
    done();
}, API_START_TIMEOUT);


describe(`API: ${BASE_URL}`, () => {
    /** **************************************************************************
     * GET MANY
     ************************************************************************** */
    test('GET MANY: SNAPSHOT OK', async () => {
        client.setUserToken();

        const response = await client.callGetMany();

        expect(response.body).toMatchSnapshot();
        expect(response.status).toBe(StatusCode.Ok);
    });

    [
        ['filter by genreId', bookQueryFactory({genreId: 1}), 2],
        ['filter by search', bookQueryFactory({search: 'Hadodrak'}), 1],
    ].forEach(([label, query, totalItems]) => {
        test(`GET MANY: ${label}`, async () => {
            client.setUserToken();

            const response = await client.callGetMany({queryParams: query});

            expect(response.body.metadata.totalItems).toEqual(totalItems);
        });
    })

    test('GET MANY: UNAUTHORIZED', async () => {
        const response = await client.callGetMany();

        expect(response.status).toBe(StatusCode.Unauthorized);
    });

    /** **************************************************************************
     * GET ONE
     ************************************************************************** */
    test('GET ONE: SNAPSHOT OK', async () => {
        client.setUserToken();

        const response = await client.callGetOne(QUERY_BOOK_ID);

        expect(response.body).toMatchSnapshot();
        expect(response.status).toBe(StatusCode.Ok);
    });

    test('GET ONE: UNAUTHORIZED', async () => {
        const response = await client.callGetOne(QUERY_BOOK_ID);

        expect(response.status).toBe(StatusCode.Unauthorized);
    });

    test('GET ONE: NOT FOUND', async () => {
        client.setUserToken();

        const response = await client.callGetOne(INVALID_BOOK_ID);

        expect(response.status).toBe(StatusCode.NotFound);
    });

    /** **************************************************************************
     * CREATE
     ************************************************************************** */
    test('CREATE: SNAPSHOT OK', async () => {
        jest.spyOn(IsbnApiService.prototype, 'getBookInfo')
            .mockImplementation(() => Promise.resolve(mockedIsbnApiResponse));
        client.setAdminToken();

        const response = await client.callPost({
            data: createBookFactory({isbn: '978-3-319-25557-7'})
        });

        expect(response.body).toMatchSnapshot();
        expect(response.status).toBe(StatusCode.Ok);
    });

    [
        ['name null', createBookFactory({name: null})],
        ['name too long', createBookFactory({name: generateString(51)})],
        ['isbn null', createBookFactory({isbn: null})],
        ['isbn too long', createBookFactory({isbn: generateString(18)})],
    ].forEach(([label, data]) => {
        test(`CREATE: BAD REQUEST, error: ${label}`, async () => {
            client.setAdminToken();

            const response = await client.callPost({data});

            expect(response.body.name).toEqual(ValidationFailed.name);
            expect(response.status).toBe(StatusCode.BadRequest);
        });
    });

    test('CREATE: BAD REQUEST, error: isbn not unique', async () => {
        client.setAdminToken();
        const data = createBookFactory({isbn: '978-3-16-148410-0'})

        const response = await client.callPost({data});

        expect(response.body.name).toEqual(IsbnNotUnique.name);
        expect(response.status).toBe(StatusCode.BadRequest);
    });

    test('CREATE: BAD REQUEST, error: invalid genre', async () => {
        client.setAdminToken();
        const data = createBookFactory({genreId: 4654})

        const response = await client.callPost({data});

        expect(response.body.name).toEqual(InvalidGender.name);
        expect(response.status).toBe(StatusCode.BadRequest);
    });

    test('CREATE: UNAUTHORIZED', async () => {
        const response = await client.callPost({data: createBookFactory()});

        expect(response.status).toBe(StatusCode.Unauthorized);
    });

    test('CREATE: FORBIDDEN', async () => {
        client.setUserToken();

        const response = await client.callPost({data: createBookFactory()});

        expect(response.status).toBe(StatusCode.Forbidden);
    });

    /** **************************************************************************
     * UPDATE
     ************************************************************************** */
    test('UPDATE: SNAPSHOT OK', async () => {
        client.setAdminToken();

        const response = await client.callPut(EDIT_BOOK_ID, {data: updateBookFactory()});

        expect(response.body).toMatchSnapshot();
        expect(response.status).toBe(StatusCode.Ok);
    });

    [
        ['name null', createBookFactory({name: null})],
        ['name too long', createBookFactory({name: generateString(51)})]
    ].forEach(([label, data]) => {
        test(`UPDATE: BAD REQUEST, error: ${label}`, async () => {
            client.setAdminToken();

            const response = await client.callPut(EDIT_BOOK_ID, {data});

            expect(response.body.name).toEqual(ValidationFailed.name);
            expect(response.status).toBe(StatusCode.BadRequest);
        });
    });

    test('UPDATE: BAD REQUEST, error: invalid genre', async () => {
        client.setAdminToken();
        const data = updateBookFactory({genreId: 4654})

        const response = await client.callPut(EDIT_BOOK_ID, {data});

        expect(response.body.name).toEqual(InvalidGender.name);
        expect(response.status).toBe(StatusCode.BadRequest);
    });

    test('UPDATE: UNAUTHORIZED', async () => {
        const response = await client.callPut(EDIT_BOOK_ID, {data: updateBookFactory()});

        expect(response.status).toBe(StatusCode.Unauthorized);
    });

    test('UPDATE: FORBIDDEN', async () => {
        client.setUserToken();

        const response = await client.callPut(EDIT_BOOK_ID, {data: updateBookFactory()});

        expect(response.status).toBe(StatusCode.Forbidden);
    });

    test('UPDATE: NOT FOUND', async () => {
        client.setAdminToken();

        const response = await client.callPut(INVALID_BOOK_ID, {data: updateBookFactory()});

        expect(response.status).toBe(StatusCode.NotFound);
    });

    /** **************************************************************************
     * DELETE
     ************************************************************************** */
    test('DELETE: SNAPSHOT OK', async () => {
        client.setAdminToken();

        const response = await client.callDelete(DELETE_BOOK_ID);

        expect(response.body).toMatchSnapshot();
        expect(response.status).toBe(StatusCode.Ok);
    });

    test('DELETE: UNAUTHORIZED', async () => {
        const response = await client.callDelete(DELETE_BOOK_ID);

        expect(response.status).toBe(StatusCode.Unauthorized);
    });

    test('DELETE: FORBIDDEN', async () => {
        client.setUserToken();

        const response = await client.callDelete(DELETE_BOOK_ID);

        expect(response.status).toBe(StatusCode.Forbidden);
    });

    test('DELETE: NOT FOUND', async () => {
        client.setAdminToken();

        const response = await client.callDelete(INVALID_BOOK_ID);

        expect(response.status).toBe(StatusCode.NotFound);
    });
});

afterEach(done => {
    jest.restoreAllMocks();
    client.clearToken();
    done();
});


afterAll(async done => {
    await client.dispose();
    done();
});
