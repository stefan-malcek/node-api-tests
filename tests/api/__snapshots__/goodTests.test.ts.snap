// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API: /api/books CREATE: SNAPSHOT OK 1`] = `
Object {
  "data": Object {
    "created": "2020-05-01T10:42:56.784Z",
    "genre": Object {
      "id": 1,
      "name": "Science fiction",
    },
    "id": 2,
    "isbn": "978-3-319-25557-7",
    "name": "I'll do it tomorrow",
    "url": "https://we.did.it",
  },
  "message": "Ok",
  "metadata": null,
  "statusCode": 200,
}
`;

exports[`API: /api/books DELETE: SNAPSHOT OK 1`] = `
Object {
  "message": "Ok",
  "metadata": null,
  "statusCode": 200,
}
`;

exports[`API: /api/books GET MANY: SNAPSHOT OK 1`] = `
Object {
  "data": Array [
    Object {
      "created": "2020-02-28T14:45:21.464Z",
      "genre": Object {
        "id": 1,
        "name": "Science fiction",
      },
      "id": 1,
      "isbn": "978-3-16-148410-0",
      "name": "The mighty Hadodrak.",
    },
    Object {
      "created": "2020-03-28T14:12:21.444Z",
      "genre": Object {
        "id": 5,
        "name": "Detective",
      },
      "id": 2,
      "isbn": "978-3-16-148410-1",
      "name": "Who took greek yogurt from the fridge?",
    },
    Object {
      "created": "2020-04-28T12:45:21.214Z",
      "genre": Object {
        "id": 4,
        "name": "Horror",
      },
      "id": 3,
      "isbn": "978-3-16-148410-2",
      "name": "API testing nightmare.",
    },
  ],
  "message": "Ok",
  "metadata": Object {
    "page": 1,
    "pageSize": 3,
    "totalItems": 4,
  },
  "statusCode": 200,
}
`;

exports[`API: /api/books GET ONE: SNAPSHOT OK 1`] = `
Object {
  "data": Object {
    "created": "2020-04-28T12:45:21.214Z",
    "genre": Object {
      "id": 4,
      "name": "Horror",
    },
    "id": 3,
    "isbn": "978-3-16-148410-2",
    "name": "API testing nightmare.",
  },
  "message": "Ok",
  "metadata": null,
  "statusCode": 200,
}
`;

exports[`API: /api/books UPDATE: SNAPSHOT OK 1`] = `
Object {
  "data": Object {
    "created": null,
    "genre": Object {
      "id": 4,
      "name": "Romance",
    },
    "id": 4,
    "isbn": "878-3-16-148410-2",
    "lastModified": "2020-05-01T10:42:56.784Z",
    "name": "Friday night",
  },
  "message": "Ok",
  "metadata": null,
  "statusCode": 200,
}
`;
